tidy3d/plugins/sweep/method.py

```diff
@@ -0,0 +1,212 @@
+        fn_post: Callable,
+        path_dir: str = None,
+        **batch_kwargs,
+    ) -> Tuple[Any]:
+        """Defines the search algorithm (batched)."""
+
+        # get all function inputs
+        fn_args, num_points = self._assemble_args(design_variables)
+
+        def get_task_name(pt_index: int, sim_index: int, fn_kwargs: dict) -> str:
+            """Get task name for 'index'-th set of function kwargs."""
+            try:
+                kwarg_str = str(fn_kwargs)
+                if sim_index is not None:
+                    return f"{kwarg_str}_{sim_index}"
+                return kwarg_str
+            # just to be safe, handle the case if this doesnt work
+            except ValueError:
+                return f"{pt_index}_{sim_index}"
+
+        # for each point, construct the simulation inputs into a dict
+        simulations = {}
+        task_name_mappings = []
+        for i in range(num_points):
+            fn_kwargs = {key: vals[i] for key, vals in fn_args.items()}
+            sim = fn_pre(**fn_kwargs)
+            if isinstance(sim, Simulation):
+                task_name = get_task_name(pt_index=i, sim_index=None, fn_kwargs=fn_kwargs)
+                simulations[task_name] = sim
+                task_name_mappings.append([task_name])
+            elif isinstance(sim, dict):
+                task_name_mappings.append({})
+                for name, _sim in sim.items():
+                    task_name = get_task_name(pt_index=1, sim_index=name, fn_kwargs=fn_kwargs)
+                    simulations[task_name] = _sim
+                    task_name_mappings[i][name] = task_name
+            else:
+                task_name_mappings.append([])
+                for j, _sim in enumerate(sim):
+                    task_name = get_task_name(pt_index=i, sim_index=j, fn_kwargs=fn_kwargs)
+                    simulations[task_name] = _sim
+                    task_name_mappings[i].append(task_name)
+
+        # run in batch
+        batch_data = self._run_batch(simulations=simulations, path_dir=path_dir, **batch_kwargs)
+        task_id_dict = batch_data.task_ids
+
+        # run post processing on each data
+        result = []
+        task_ids = []
+        for task_names_i in task_name_mappings:
+            if isinstance(task_names_i, dict):
+                task_ids.append([task_id_dict[task_name] for task_name in task_names_i.values()])
+                kwargs_dict = {
+                    kwarg_name: batch_data[task_name]
+                    for kwarg_name, task_name in task_names_i.items()
+                }
+                val = fn_post(**kwargs_dict)
+            else:
+                task_ids.append([task_id_dict[task_name] for task_name in task_names_i])
+                args_list = (batch_data[task_name] for task_name in task_names_i)
+                val = fn_post(*args_list)
+
+            result.append(val)
+
+        return fn_args, result, task_ids
+
+
+class MethodRandom(MethodIndependent):
```

Comment:
Sampling the variable independently usually leads to very poor MonteCarlo performance, which I guess would be one of the main uses for this plugin. I'd much prefer if we could use something like Latin Hypercube or orthogonal sampling, maybe through (https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.qmc.LatinHypercube.html or some other implementation).

Reply:
I'm in favor of this, but am not sure about how to implement it for discrete sets (like integers or VariableAny). Any idea? 

Do you think this warrants its own Method? eg. MethodMonteCarlo?

Reply:
According to the docs in the scipy implementation, it's safe to scale and take the floor of the random number generated in the [0, 1) range (w.r.t. maintaining distribution characteristics).

Reply:
> 1. Should `MethodRandom` just use this `scipy.qmc.LatinHypercube` approach? Or would it make sense to make a separate class for Monte Carlo Methods?

Honestly, I don't see a reason to have just independently random picks. I'd just use LH sampling and note it in the documentation. The average user most likely doesn't care and is not aware of the difference. The user who understands it is probably knowledgeable enough to insert their own variable transformation if they want to reshape the distribution or generate the samples themselves and use `MethodIndependent` (can it be made a concrete class that just runs the given values?)

> 2. Would it make sense to try to generalize the monte carlo methods, for example to use any of the `spicy.qmc` engines [here](https://docs.scipy.org/doc/scipy/reference/stats.qmc.html)? Or other samplers. To what extent do you think we should generalize these methods? If I make this one `MethodRandom`, it might limit somewhat what we can do later.

This might actually be a better approach: having `MethodRandom` accept a scipy.stats.qmc.QMCEngine with LH as the default. This would enable advanced users to to whatever they wish, but a good default for everyone else.

Reply:
In fact, it doesn't *have* to be a `QMCEngine` object, but any instance that provides a `.random` method with the desired signature.

Reply:
sounds good. So for integers, I think scaling the distribution and binning makes sense, but for selecting non-numeric objects (like VariablAny) think it's fine to use the same approach? Like does the Monte Carlo algorithm here apply?

Reply:
Yes, same thing. I don't think there is any "succession" property that would make categorical values (`VariableAny`) different from an integer set.